### ISSUE TITLE
Introduce coding standards enforcement

### ISSUE THREAD

**Issue by arturponinski** [petl-developers/petl#597]
## Overview
There is a number of mature tools that could, or even should be used to enforce coding standards for the code base. 

The tools to consider are:
* pyupgrade
* flake8
* black
* isort
The configuration for those tools should be set up in the repository.

## Task list
- [ ] agree on tools to use and their configuration
- [ ] add a separate requirements file
- [ ] add a pre-commit hook that can be used by contributors
- [ ] add github action that will verify if the PR conforms the standards
- [ ] update contributing section of the docs

Since this will affect whole codebase, it should be considered for 2.0 release

**Comment by juarezr** [comment]
I agree. Additionally we should consider:

- [pre-commit](https://pre-commit.com/) hooks
- Unifying configuration in [pyproject.toml](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#pyproject-toml)pyproject.toml)
- Reformat and lint all code
- Blocking PR on formatting / linting issues ( Codacity, CodeQL checks )

**Comment by juarezr** [comment]
- Interesting approach. Looks pretty good to me.
- My 2 cents:
  - The only thing that we need to pay attention to is breaking compatibility.
    - Ideally, we should batch all compatibility breaking changes in v2.0
    - But as we are volunteers, we can postpone anything beyond our availability
    - As everything is working, we don't have a strong reason to tag a v2.0 right now.
    - Maybe typing would be such a killer feature to drop compatibility...
  - Using a formatted v1.8.x in parallel with a non-release v2.0.x branch would be amazing
  - Similarly, we can do v1.9.x as bug fix releases after we release v2.0.x

**Comment by juarezr** [comment]
@arturponinski ,

Alter #613 I started wondering if:

- There are other places in the source code that would benefit from compliance with PEP 479
- There are other PEP-XXX modernization efforts that would benefit `petl`
- Should we keep track of these kinds of improvements

**Comment by juarezr** [comment]
# New Python Tools

Nowadays, the tooling around Python has improved with recent tools like [Ruff](https://github.com/astral-sh/ruff), and [uv](https://astral.sh/blog/uv).

## Ruff

### About Ruff

From the Ruff repository cover:

> An extremely fast Python linter and code formatter, written in Rust.
> Ruff aims to be orders of magnitude faster than alternative tools while integrating more functionality behind a single, common interface.
>
> Ruff can be used to replace [Flake8](https://pypi.org/project/flake8/) (plus dozens of plugins), [Black](https://github.com/psf/black), [isort](https://pypi.org/project/isort/), [pydocstyle](https://pypi.org/project/pydocstyle/), [pyupgrade](https://pypi.org/project/pyupgrade/), [autoflake](https://pypi.org/project/autoflake/), and more, all while executing tens or hundreds of times faster than any individual tool.
>
> Ruff is extremely actively developed and used in major open-source projects like:
    [Apache Airflow](https://github.com/apache/airflow)
    [Apache Superset](https://github.com/apache/superset)
    [FastAPI](https://github.com/tiangolo/fastapi)
    [Hugging Face](https://github.com/huggingface/transformers)
    [Pandas](https://github.com/pandas-dev/pandas)
    [SciPy](https://github.com/scipy/scipy)
   ...and [many more](https://github.com/astral-sh/ruff#whos-using-ruff).

### Thoughts on Ruff

It would be interesting to investigate the advantages/disadvantages of settling on a single tool like Ruff and optionally supporting the traditional ones like Flake8, Black, and pydocstyle.

As we plan to do a full code reformatting to further check style on a commit/pull request basis, we will need to have a consensus around which tools/defaults will be used.

In my experience, **Ruff** checks all boxes here.

## uv

### About uv

From the `uv` [repository cover](https://github.com/astral-sh/uv?tab=readme-ov-file#uv):

> An extremely fast Python package and project manager, written in Rust.
Highlights
>
> 🚀 A single tool to replace pip, pip-tools, pipx, poetry, pyenv, twine, virtualenv, and more.
> ⚡️ [10-100x faster](https://github.com/astral-sh/uv/blob/main/BENCHMARKS.md) than pip.
> 🗂️ Provides [comprehensive project management](https://github.com/astral-sh/uv#projects), with a [universal lockfile](https://docs.astral.sh/uv/concepts/projects/layout#the-lockfile).
> ❇️ [Runs scripts](https://github.com/astral-sh/uv#scripts), with support for [inline dependency metadata](https://docs.astral.sh/uv/guides/scripts#declaring-script-dependencies).
> 🐍 [Installs and manages](https://github.com/astral-sh/uv#python-versions) Python versions.
> 🛠️ [Runs and installs](https://github.com/astral-sh/uv#tools) tools published as Python packages.
> 🔩 Includes a [pip-compatible interface](https://github.com/astral-sh/uv#the-pip-interface) for a performance boost with a familiar CLI.
> 🏢 Supports Cargo-style [workspaces](https://docs.astral.sh/uv/concepts/projects/workspaces) for scalable projects.
> 💾 Disk-space efficient, with a [global cache](https://docs.astral.sh/uv/concepts/cache) for dependency deduplication.
> ⏬ Installable without Rust or Python via curl or pip.
> 🖥️ Supports macOS, Linux, and Windows.
>
> uv is backed by [Astral](https://astral.sh/), the creators of [Ruff](https://github.com/astral-sh/ruff).

Also check this interesting [blog post](https://astral.sh/blog/uv) about `uv` advantages:

> uv represents an intermediary milestone in our pursuit of a ["Cargo for Python"](https://blog.rust-lang.org/2016/05/05/cargo-pillars.html#pillars-of-cargo): a unified Python package and project manager that is extremely fast, reliable, and easy to use.
>
>Think: a single binary that bootstraps your Python installation and gives you everything you need to be productive with Python, bundling not only pip, pip-tools, and virtualenv, but also pipx, tox, poetry, pyenv, ruff, and more.

### Thoughts on uv

It would be interesting to investigate the use of [uv](https://astral.sh/blog/uv) for establishing a default developer workflow for the `petl` repository. 

Perhaps it would be an interesting option for using in the CI/CD jobs/workflows in the Git Actions we currently use.

Even if we decide on using [uv](https://astral.sh/blog/uv) as the default workflow, it would be good to continue supporting traditional tools like `pip`, `pip-tools`, `virtualenv`, `virtualenvwrapper`, and `tox` as they are the most used.

Also, it would be possible to support some optional workflows around other package managers like [conda](https://docs.conda.io), or [poetry](https://python-poetry.org/) based on contributions.

Notice that the choice of developer workflow or package manager should not impact neither the CI/CD workflows/jobs/checks, neither the release to the [PyPi](https://pypi.org/project/petl/) package index and the [Anaconda](https://anaconda.org/conda-forge/petl) distribution.
